src/pykeen/datasets/analysis.py

```diff
@@ -0,0 +1,484 @@
+            continue
+        for r, ht in pairs.items():
+            confidence = len(lhs.intersection(ht)) / support
+            yield PatternMatch(r, "composition", support, confidence)
+
+
+def _determine_patterns(
+    mapped_triples_list: Collection[Tuple[int, int, int]],
+) -> Iterable[PatternMatch]:
+    # indexing triples for fast lookup of entity pair sets
+    pairs: DefaultDict[int, Set[Tuple[int, int]]] = defaultdict(set)
+    for h, r, t in mapped_triples_list:
+        pairs[r].add((h, t))
+    # unary
+    yield from yield_unary_patterns(pairs=pairs)
+
+    # binary
+    yield from yield_binary_patterns(pairs=pairs)
+
+    # ternary
+    yield from yield_ternary_patterns(mapped_triples_list, pairs)
+
+
+def relation_classification(
+    dataset: Dataset,
+    min_support: int = 0,
+    min_confidence: float = 0.95,
+    drop_confidence: bool = True,
+    parts: Optional[Collection[str]] = None,
+) -> pandas.DataFrame:
+    r"""
+    Categorize relations based on patterns from RotatE [sun2019]_.
+
+    The relation classifications are based upon checking whether the corresponding rules hold with sufficient support
+    and confidence. By default, we do not require a minimum support, however, a relatively high confidence.
+
+    The following four non-exclusive classes for relations are considered:
+
+    - symmetry
+    - anti-symmetry
+    - inversion
+    - composition
+
+    This method generally follows the terminology of association rule mining. The patterns are expressed as
+
+    .. math ::
+
+        X_1 \land \cdot \land X_k \implies Y
+
+    where $X_i$ is of the form $r_i(h_i, t_i)$, and some of the $h_i / t_i$ might re-occur in other atoms.
+    The *support* of a pattern is the number of distinct instantiations of all variables for the left hand side.
+    The *confidence* is the proportion of these instantiations where the right-hand side is also true.
+
+    :param dataset:
+        The dataset to investigate.
+    :param min_support:
+        A minimum support for patterns.
+    :param min_confidence:
+        A minimum confidence for the tested patterns.
+    :param drop_confidence:
+        Whether to drop the support/confidence information from the result frame, and also drop duplicates.
+    :param parts:
+        Only use certain parts of the dataset, e.g., train triples. Defaults to using all triples, i.e.
+        {"training", "validation", "testing}.
+
+    .. warning ::
+        If you intend to use the relation categorization as input to your model, or hyperparameter selection, do *not*
+        include testing triples to avoid leakage!
+
+    :return:
+        A dataframe with columns {"relation_id", "pattern", "support"?, "confidence"?}.
+    """
+    # normalize parts
+    if parts is None:
+        parts = dataset.factory_dict.keys()
+    parts = [parts] if isinstance(parts, str) else parts
+
+    # include part hash into cache-file name
+    part_hash = hashlib.sha512("".join(sorted(parts)))[:16]
```

Comment:
alternate concept for part_hash:

```python
str(int('training' in parts)) + str(int('testing' in parts)) + str(int('validation' in parts))
```

this would just be a string of 0/1 three times in a row

Reply:
The hash is not good enough yet, since it does not cover different split sizes :confused:

Reply:
ooooh interesting, maybe we need a generalize hash for a given dataset that represents which triples ended up in testing, training, and validation?

Reply:
For now I went with

```python
"".join(map(str, sorted(mapped_triples.tolist()))).encode("utf8")
```